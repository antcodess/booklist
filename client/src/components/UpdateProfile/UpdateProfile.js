import React, { useState } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { createStructuredSelector } from 'reselect';

import FormInput from '../FormInput/FormInput';
import CustomButton from '../CustomButton/CustomButton';

import { selectCurrentUser } from '../../redux/user/user.selectors';
import { updateNameStart, updateEmailStart, updatePasswordStart } from '../../redux/user/user.actions';
import { setAlert } from '../../redux/alert/alert.actions';

import './UpdateProfile.scss';

export const UpdateProfile = ({ currentUser, updateNameStart, updateEmailStart, updatePasswordStart, setAlert }) => {

  const [userCredentials, setUserCredentials] = useState({
    displayName: currentUser.displayName,
    email: currentUser.email,
    emailInputPassword: '',
    oldPassword: '',
    newPassword: '',
    confirmNewPassword: ''
  });

  const [emailPasswordInputHidden, setEmailPasswordInputHidden] = useState(true);

  const { displayName, 
          email, 
          oldPassword,
          newPassword,
          confirmNewPassword,
          emailInputPassword } = userCredentials;

  const onInputChange = (event) => {
    const { name, value } = event.target;
    setUserCredentials({ ...userCredentials, [name]: value })
  };

  const onEmailInputClick = () => {
    setEmailPasswordInputHidden(false);
  };

  const onEmailButtonClick = () => {
    setEmailPasswordInputHidden(false)
    if (!emailInputPassword) {
      setAlert({
        status: 'regular',
        message: 'Please enter your password.'
      });
      return;
    }
  };

  const onNameSubmit = (event) => {
    event.preventDefault();
    updateNameStart(displayName);
  };

  const onEmailSubmit = (event) => {
    event.preventDefault();
    if (!emailInputPassword) {
      setAlert({
        status: 'failure',
        message: 'Please enter your password.'
      });
      return;
    }
    updateEmailStart(email, emailInputPassword);
  };

  const onPasswordSubmit = (event) => {
    event.preventDefault();
    if (newPassword !== confirmNewPassword) {
      setAlert({
        status: 'failure',
        message: 'New passwords must match.'
      });
      return;
    }
    updatePasswordStart(oldPassword, newPassword);
  };


  return (
    <section className='update-profile'>
        <div className='name-email'>
        <form onSubmit={onNameSubmit}>
          <FormInput 
            name='displayName'
            type='name'
            value={displayName}
            onChange={onInputChange} 
            label='Name'
            placeholder='Name' 
            required 
          />
          <CustomButton>Update Name</CustomButton>
        </form>
        <form onSubmit={onEmailSubmit}>
          <FormInput 
            name='email'
            type='email'
            value={email}
            onChange={onInputChange}
            onClick={onEmailInputClick} 
            label='Email'
            placeholder='Email' 
            required 
          />

          { !emailPasswordInputHidden ? 
            <FormInput 
              name='emailInputPassword'
              type='password'
              value={emailInputPassword}
              onChange={onInputChange} 
              label='Password'
              placeholder='Enter Password' 
              required 
            />
            : null
          }
          <CustomButton onClick={onEmailButtonClick}>Update Email</CustomButton>
        </form>
        </div>
        <form className='update-password' onSubmit={onPasswordSubmit}>
          <FormInput 
            name='oldPassword'
            type='password'
            value={oldPassword}
            onChange={onInputChange} 
            label='Old Password'
            placeholder='Old Password' 
            required 
          />
          <FormInput 
            name='newPassword'
            type='password'
            value={newPassword}
            onChange={onInputChange} 
            label='New Password'
            placeholder='New Password' 
            required 
          />
          <FormInput 
            name='confirmNewPassword'
            type='password'
            value={confirmNewPassword}
            onChange={onInputChange} 
            label='Confirm New Password'
            placeholder='Confirm New Password' 
            required 
          />
          <CustomButton>Update Password</CustomButton>
        </form>
    </section>
  )
}

const mapStateToProps = createStructuredSelector({
  currentUser: selectCurrentUser
});

export const mapDispatchToProps = (dispatch, ownProps) => ({
  updateNameStart: displayName => dispatch(updateNameStart(displayName, ownProps)),
  updateEmailStart: (email, password) => dispatch(updateEmailStart(email, password, ownProps)),
  updatePasswordStart: (oldPassword, newPassword) => dispatch(updatePasswordStart(oldPassword, newPassword, ownProps)),
  setAlert: alert => dispatch(setAlert(alert))
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(UpdateProfile));