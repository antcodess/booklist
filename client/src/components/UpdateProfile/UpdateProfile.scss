.update-profile {

  .name-email {
    display: flex;
    justify-content: space-between;
  }
  span {
    text-align: left;
    cursor: pointer;

    &:hover {
      opacity: 0.7;
    }
  }
  
  form {
    margin: 10px 0px 30px;
    width: 49%;

    .custom-button {
      margin: 0 auto;
      width: 100%;
    }
  }
  .update-password {
    width: 100%;
  }
}