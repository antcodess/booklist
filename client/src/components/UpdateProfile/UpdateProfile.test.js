import React from 'react';
import { shallow } from 'enzyme';
import { UpdateProfile } from './UpdateProfile';
import { mapDispatchToProps } from './UpdateProfile';

describe('UpdateProfile Component', () => {
  let wrapper;
  let mockCurrentUser;
  let mockUpdateNameStart;
  let mockUpdateEmailStart;
  let mockUpdatePasswordStart;
  let mockSetAlert;

  const mockUpdateNamePayload = {
    type: 'UPDATE_NAME_START',
    payload: {
      displayName: 'Sample Name',
      ownProps: undefined
    }
  };

  const mockUpdateEmailStartPayload = {
    type: 'UPDATE_EMAIL_START',
    payload: {
      email: 'email@example.com',
      password: 'password',
      ownProps: undefined
    }
  };

  const mockUpdatePasswordStartPayload = {
    type: 'UPDATE_PASSWORD_START',
    payload: {
      oldPassword: 'oldpassword',
      newPassword: 'newpassword',
      ownProps: undefined
    }
  };

  const mockUserCredentials = {
    displayName: 'Sample Name',
    email: 'email@example.com', 
    password: 'password',
    oldPassword: 'oldpassword',
    newPassword: 'newpassword',
    confirmNewPassword: 'newpassword'
  };

  const mockAlert = {
    status: 'failure',
    message: 'Passwords must match.'
  };

  const mockSetAlertPayload = {
    type: 'SET_ALERT',
    payload: {
      status: 'failure',
      message: 'Passwords must match.'
    }
  };

  const ownProps = {};
  
  beforeEach(() => {
    
    mockCurrentUser = { id: 1, name: 'Sample User'};
    mockUpdateNameStart = jest.fn();
    mockUpdateEmailStart = jest.fn();
    mockUpdatePasswordStart = jest.fn();
    mockSetAlert = jest.fn();

    const mockProps = {
      currentUser: mockCurrentUser, 
      updateNameStart: mockUpdateNameStart, 
      updateEmailStart: mockUpdateEmailStart, 
      updatePasswordStart: mockUpdatePasswordStart, 
      setAlert: mockSetAlert,
      userCredentials: {
        displayName: 'Sample Name',
        email: 'email@example.com', 
        emailInputPassword: '',
        oldPassword: 'oldpassword',
        newPassword: 'newpassword',
        confirmNewPassword: 'newpassword'
      }
    };
    
    wrapper = shallow(<UpdateProfile { ...mockProps } />);
  });

  it('should render UpdateProfile Component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should dispatch updateNameStart action when clicked', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).updateNameStart(mockUserCredentials.displayName, ownProps);
    expect(dispatch.mock.calls[0][0]).toEqual(mockUpdateNamePayload);
  });

  it('should dispatch updateEmailStart action when clicked', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).updateEmailStart(
      mockUserCredentials.email,
      mockUserCredentials.password,  
      ownProps
    );
    expect(dispatch.mock.calls[0][0]).toEqual(mockUpdateEmailStartPayload);
  });

  it('should dispatch updatePasswordStart action when clicked', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).updatePasswordStart(
      mockUserCredentials.oldPassword,
      mockUserCredentials.newPassword,  
      ownProps
    );
    expect(dispatch.mock.calls[0][0]).toEqual(mockUpdatePasswordStartPayload);
  });

  it('should dispatch setAlert action when clicked', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).setAlert(mockAlert);
    expect(dispatch.mock.calls[0][0]).toEqual(mockSetAlertPayload);
  });
});
