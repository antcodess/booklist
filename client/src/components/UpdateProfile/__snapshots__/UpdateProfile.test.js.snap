// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateProfile Component should render UpdateProfile Component 1`] = `ShallowWrapper {}`;
